crates/polars-lazy/src/scan/file_list_reader.rs

```diff
@@ -1,73 +1,176 @@
+            MultipleReaderFactories::SpecificPaths { paths } => {
+                write!(f, "[")?;
+                for path in paths {
+                    write!(f, "{}, ", path.to_string_lossy())?;
+                }
+                write!(f, "]")
+            },
+        }
+    }
+}
+
+impl MultipleReaderFactories {
+    /// Create a new ReaderSources with a specific glob path.
+    pub fn new_glob(path: impl AsRef<Path>) -> Self {
+        Self::GlobPath { path: path.as_ref().to_owned() }
+    }
+
+    /// Create a new ReaderSources with a set of specific, non-glob paths.
+    pub fn new_paths(paths: Vec<PathBuf>) -> Self {
+        Self::SpecificPaths { paths }
+    }
+
+    /// Get list of specific reader factories.
+    ///
+    /// Returns [None] if path is not a glob pattern.
+    fn iter_readers(&self, cloud_options: Option<&CloudOptions>) -> PolarsResult<ReaderIterator> {
+        match self {
+            MultipleReaderFactories::GlobPath { path } => {
+                let path_str = path.to_string_lossy();
+                polars_glob(&path_str, cloud_options)
+            },
+            MultipleReaderFactories::SpecificPaths { paths } => {
+                Ok(Box::new(paths.clone().into_iter().map(|path| {
+                    let path_str = path.to_string_lossy();
+                    if is_cloud_url(path_str.as_ref()) {
+                        Ok(ReaderFactory::Remote {
+                            location: path_str.to_string(),
+                        })
+                    } else {
+                        let f = File::open(path)?;
+                        Ok(ReaderFactory::Local {
+                            path,
+                            reader: Box::new(f),
+                        })
+                    }
+                })))
+            },
+        }
+        // if paths.is_empty() {
+        //     let path_str = self.path().to_string_lossy();
+        //     if path_str.contains('*') || path_str.contains('?') || path_str.contains('[') {
+        //         polars_glob(&path_str, self.cloud_options()).map(Some)
+        //     } else {
+        //         Ok(None)
+        //     }
+        // } else {
+        //     polars_ensure!(self.path().to_string_lossy() == "", InvalidOperation: "expected only a single path argument");
+        //     // Lint is incorrect as we need static lifetime.
+        //     #[allow(clippy::unnecessary_to_owned)]
+        //     Ok(Some(Box::new(paths.to_vec().into_iter().map(Ok))))
+        // }
+    }
+}
+/// A LazyFileListReader can be in one of two states: either it has a set of
+/// sources (a filesystem glob, or a list of file paths/uris it needs to iterate
+/// over), or it has a single, specific ReaderFactory. This isn't a great design
+/// and there should really be separate structs for the two states, e.g. a
+/// builder and a final state.
+#[derive(Clone)]
+pub(crate) enum SpecificOrMultipleReaderFactories {
```

Comment:
I feel we are overcomplicating now. We end up with 3 nested enums. 

I understand that we can disambiguate and factor out some logic. But I rather have a single `enum` that represents a state for a file. And if we have multiple files, we store that single enum in a vec.

Or we add all states in single enum. Not entirely sure, but nesting the state feel to overengineered at the moment as we are not sure here yet about the final state.